### ISSUE TITLE
1.12 unit filter formula aborts wesnoth if there is an expression error

### ISSUE THREAD

**Issue by Shrieker1** [wesnoth/wesnoth#1946]
Wesnoth 1.12.6, (I don't have 1.13 available to test if this is fixed)

I made an error trying to use double equals sign to compare equality in an event unit filter formula like following

~~~~
[event]
  name=moveto
  [filter]
    formula="'$player[$side_number].race' == 'humans'"
  [/filter]
~~~~

Too bad this resulted to wesnoth aborting instead of just error message to log and failing to handle the event with broken filter.

Wesnoth log:
~~~~
Formula error in formula:1
In formula = 
Error: Illegal unary operator: '='

Game will be aborted.
~~~~

**Comment by gfgtdf** [comment]
probably effects 1.13 aswell, see this TODO: https://github.com/wesnoth/wesnoth/blob/master/src/units/filter.cpp#L562

**Comment by CelticMinstrel** [comment]
By the way, it's better where possible to use the WFL variables rather than the WML variables. If you want the event to trigger for any human unit, `formula="race = 'humans'"` should do the trick, or if you prefer to be a little more explicit, `formula="unit.race = 'humans'"`. Mind you, this is assuming that `$player[$side_number]` always resolves to the unit being matched by the filter, but if it doesn't, then I'd be questioning the filter for other reasons.

**Comment by Shrieker1** [comment]
Obviously that is minimal test case only. There is much more than just that formula in my code. But with my filter testing actual bug would be much harder than that minimal test which would trigger the bug if dropping it to any scenario (for example using lua command mode)

player[side_nubmer].race evaluates to side faction. The moveto handler is implemented as faction specific with minor variations. I'm now working on merging similar faction specifics handlers to one generic. I started conversion from the code part that made units look same to filters forcing me to add a quick formula based faction check until that part of code has been refactored to generic form.

But for a few other cases I find it that checking scenario variables in formulas is the simplest way to implement some filters which should behaviour differently depending on state variables. If I used separate variable conditionals to check scenario variables I would double or triple branches in some filters.
